media/windows/ECC-MQTT-IoT/ECC_MQTT_IoT_SQLite.py

```diff
@@ -101,16 +124,149 @@
+        # used.  If more than (backupCount) files are already zipped, delete the oldest and rename
+        # the remaining to free up the last number for re-use.  This ensures that the newest gzip
+        # will always be the highest number.
+        nextgz_file_int = lastgz_file_int + 1
+        if nextgz_file_int > self.backupCount:
+            nextgz_file_int = self.backupCount
+            # Since we already have the maximum gzip'd logs, delete the oldest, then rename the
+            # remaining to free up the last slot.
+            oldest_gz_log = self.baseFilename + ".1.gz"
+            os.remove(oldest_gz_log)
+            # Rename the remaining gzip'd logs...
+            arch_files = glob.iglob(base_file_pattern)
+            for archFile in arch_files:
+                fn = archFile.split('.')
+                # The gzip'd file "number" is in array element 2
+                newgz_file_num = int(fn[2]) - 1
+                newgz_file = fn[0] + "." + fn[1] + "." + str(newgz_file_num) + ".gz"
+                os.rename(archFile, newgz_file)
+        else:
+            # haven't hit the maximum backup file count yet, so simply construct the
+            # next gzip file name based on the existing base file name components
+            fn = (self.baseFilename + ".1.gz").split('.')
+
+        nextgz_file_name = fn[0] + "." + fn[1] + "." + str(nextgz_file_int) + ".gz"
+        dest = self.baseFilename + ".1"
+
+        # Start a thread to gzip the output file so as not to delay primary task(s)
+        done_flag = False
+        ex = futures.ThreadPoolExecutor(max_workers=2)
+        print('Starting GZIP thread for log file {dest}...')  # debug
+        f = ex.submit(self.gzip_log, dest, nextgz_file_name)
+        f.add_done_callback(self.done)
+
+    @staticmethod
+    def gzip_log(dest, nextgz_file_name):
+
+        global gzip_in_progress
+
+        gzip_in_progress = True
+        try:
+            with open(dest, 'rb') as f_in, gzip.open(nextgz_file_name, 'wb') as f_out:
+                f_out.writelines(f_in)
+        except Exception as e:
+            return F'Error occurred during attempt to GZIP log {dest}:  {e}', dest
+        return F'Log file {dest} successfully GZIP''d...', dest
+
+    @staticmethod
+    def done(fn):
+        global done_flag
+        global gzip_in_progress
+
+        gzip_in_progress = False        # reset the GZIP-in-progress flag
+
+        result = fn.result()
+
+        if fn.cancelled():
+            print(F"GZIP thread cancelled")
+            # The destination file is returned as the second element of the result list
+            print(F"...log file {result[1]} should be manually GZIP'd and the source deleted.")
+            logger.error(F"GZIP thread cancelled")
+            logger.error(F"...log file {result[1]} should be manually GZIP'd and the source deleted.")
+        elif fn.done():
+            error = fn.exception()
```

Comment:
I notice that `gzip_log()` always returns a string.  Will `fn.exception()` ever return a true value?  Or does `gzip_log()` need to (re)`raise` the exception if it occurs?  (or possibly not even catch it, and let the `ThreadPoolExecutor` catch it, so that `fn.exception()` will return a true value...?

I'm didn't read the docs on `ThreadPoolExecutor`, so I'm unsure how they work.

Reply:
Here's an extract on the concurrent.futures exception object:
```
exception(timeout=None)
Return the exception raised by the call. If the call hasn’t yet completed then this method will wait up to timeout seconds. If the call hasn’t completed in timeout seconds, then a concurrent.futures.TimeoutError will be raised. timeout can be an int or float. If timeout is not specified or None, there is no limit to the wait time.

If the future is cancelled before completing then CancelledError will be raised.

If the call completed without raising, None is returned.
```
If I understand correctly, fn.result is the string return from the function.  I don't think an exception needs to be re-raised should it occur.  FWIW, I "borrowed" this code segment on using a done callback from here: https://pymotw.com/3/concurrent.futures/

Reply:
`fn.result()` returns the string: ah, gotcha.  That makes sense.

From that "futures" example, I think it is showing that you have to not catch the exception in the thread function, and instead let the ThreadPool catch it.  That's the only way that it will know that it can return a value from `fn.exception()` .  Otherwise, once you catch an exception, it's gone / it won't be visible to the outer ThreadPool -- i.e., `fn.exception()` won't know that it should return the exception that was raised.

I think you might want something like:

```python
    def gzip_log(dest, nextgz_file_name):

        global gzip_in_progress

        gzip_in_progress = True
        with open(dest, 'rb') as f_in, gzip.open(nextgz_file_name, 'wb') as f_out:
            f_out.writelines(f_in)
        return F'Log file {dest} successfully GZIP''d...', dest
```

And if an exception occurs in the `open()` or `gzip.open()` or `writelines()`, it'll exit out of `gzip_log()` and the outer ThreadPool will catch it.  That will allow the caller's `fn.exception()` to know that an exception actually occurred, and it can return the exception object.

If you catch and handle the exception in `gzip_log()`, then ThreadPool won't know that an exception occurred, and `fn.exception()` will return None or whatever it returns when it knows that the task is done but no exception occurred.

BTW, super cool use of `,` in the context manager `with` statement -- I didn't know you could do that! 🤓 

Final sidenote: I don't think you need the trailing `,dest` in the string return:

```python
# f strings will put the string equivalent of the expression
# in the string.  Also, you can use the "" to surround the 
# string and then use ' without escaping it in "GZIP'ed".
return f"Log file {dest} successfully GZIP'ed..."
```

Reply:
Okay, I understand the point about exception handling and will make the change.  RE: the return of `{dest}` from the routine...the reason I'm returning it is to have it available for the log / print messages in the `done` callback.  Is there a better way to handle this?  In general, I wasn't really comfortable using the globals for `gzip_in_progress` and `done_flag` either if this is to be used more generically as a handler, as these would need to be declared in each script that uses it as I see it.  If you have suggestions on how to better handle these state flags I'd be interested in your thoughts (even though the code works as is...)

Reply:
Ooooh, I see -- you were returning two separate values from the function -- that second `dest` wasn't an argument to `print`.  Got it.  Given that ThreadPool only lets us return one value, I might have chosen to return that as a dictionary instead of (effectively) a list, but that's personal preference / I don't think it really matters (i.e., toe-MAY-toe vs. toe-MAH-toe).